### PR TITLE
Fix infinite loop

### LINKED ISSUE #9443: Stack overflow in solc due to recursive BFS
## Description
Consider the following POC:
```solidity
library a {
    struct b {
        mapping ( uint => b ) c ;
    }
    function d( b memory) public ;

}
```

Run with `solc --bin -o poc poc.sol` with an solc built with asan and we got 
```
ASAN:DEADLYSIGNAL
=================================================================
==79623==ERROR: AddressSanitizer: stack-overflow on address 0x7fffe28e5ae8 (pc 0x7f9ea88fd509 bp 0x7fffe28e6350 sp 0x7fffe28e5ac0 T0)
    #0 0x7f9ea88fd508  (/usr/lib/x86_64-linux-gnu/libasan.so.4+0x79508)
    #1 0x56550e151170 in std::char_traits<char>::copy(char*, char const*, unsigned long) (/home/yongheng/solidity/asan/solc/solc+0x22c170)
    #2 0x56550e164241 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&) (/home/yongheng/solidity/asan/solc/solc+0x23f241)
    #3 0x56550e2d94de in solidity::util::Result<solidity::frontend::Type const*>::Result(solidity::frontend::Type const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) (/home/yongheng/solidity/asan/solc/solc+0x3b44de)
    #4 0x56550e2c1d00 in solidity::util::Result<solidity::frontend::Type const*>::Result(solidity::frontend::Type const*) (/home/yongheng/solidity/asan/solc/solc+0x39cd00)
    #5 0x56550e27a2f8 in solidity::frontend::StructType::interfaceType(bool) const (/home/yongheng/solidity/asan/solc/solc+0x3552f8)
    #6 0x56550e29eddd in solidity::frontend::MappingType::interfaceType(bool) const (/home/yongheng/solidity/asan/solc/solc+0x379ddd)
    #7 0x56550e2ab73c in solidity::frontend::StructType::interfaceType(bool) const::{lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}::operator()<solidity::util::BreadthFirstSearch& solidity::util::BreadthFirstSearch<solidity::frontend::StructDefinition const*>::run<{lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}>({lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}&&)::{lambda(solidity::frontend::StructDefinition const*)#1}>(auto, solidity::frontend::StructDefinition const*, {lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}&&) const (/home/yongheng/solidity/asan/solc/solc+0x38673c)
    #8 0x56550e2abfb0 in solidity::util::BreadthFirstSearch<solidity::frontend::StructDefinition const*>& solidity::util::BreadthFirstSearch<solidity::frontend::StructDefinition const*>::run<solidity::frontend::StructType::interfaceType(bool) const::{lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}>(solidity::frontend::StructType::interfaceType(bool) const::{lambda(solidity::frontend::StructDefinition const*, auto:1&&)#1}&&) (/home/yongheng/solidity/asan/solc/solc+0x386fb0)
....
```
## Environment

- Compiler version:Version: 0.6.12-develop.2020.7.17+commit.d808302c.Linux.g++
- Target EVM version (as per compiler settings):
- Framework/IDE (e.g. Truffle or Remix):
- EVM execution environment / backend / blockchain client:
- Operating system: Ubuntu 16

### PR DESCRIPTION
**Pull request by hrkrshnn**
Closes https://github.com/ethereum/solidity/issues/9443

It's easier to review the first commit alone.

The second commit is just indentation change. I'll squash it after the review.

### CODE REVIEW

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -2461,6 +2461,12 @@ TypeResult StructType::interfaceType(bool _inLibrary) const
 
 	TypeResult result{TypePointer{}};
 
+	if (this->recursive() && !(_inLibrary && location() == DataLocation::Storage))
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]:
```suggestion
	if (recursive() && !(_inLibrary && location() == DataLocation::Storage))
```

**Review by chriseth — DISMISSED** [review]

**Comment by chriseth** [comment]
It looks like the locations of some errors changed - can you check?

File: test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol [hunk 2]
```diff
@@ -0,0 +1,6 @@
+library a {
+    struct b {
+        mapping (uint => b) c ;
```
Comment (axic) [inline test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol:3]: Remind me, why do we allow mappings in structs?

Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol:3]: If you want to see an example of it being used, then
  https://github.com/JoinColony/colonyNetwork/blob/5cd0c042e270b665a49d6d024846a1e4bf4829fc/contracts/colony/ColonyDataTypes.sol#L274

Comment (hrkrshnn) [reply test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol:3]: However, there are way too many bugs related to allowing mappings inside structs and arrays. So this is a tricky feature.

**Comment by chriseth** [comment]
Please rebase.

**Review by chriseth — APPROVED** [review]

File: test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol (hunk 2 above)
Comment (leonardoalt) [reply test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol:3]: And these are still valid:
```
contract C {
    struct S {
        mapping (uint => uint) m;
    }
    S s1;
    mapping (uint => uint)[] a;
    function f() public {
        delete s1;
        delete a;
        a.pop();
    }
}
```

Comment (chriseth) [reply test/libsolidity/syntaxTests/structs/recursion/recursive_struct_nested_mapping_storage.sol:3]: Yes, there is an open issue about `delete` and `pop` for data structures that contain mappings.

**Pull request merged**